### PR TITLE
Fix drawing for HITL

### PR DESCRIPTION
**Pull request by nerdai**
# Description

This PR fixes drawing of Workflows that use HITL pattern. It introduces a conceptual "external_step"

image

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [x] I believe this change is already covered by existing unit tests

### CODE REVIEW

File: llama-index-core/llama_index/core/workflow/events.py [hunk 1]
```diff
@@ -132,18 +140,30 @@ def __init__(self, result: Any = None) -> None:
...
 
+class BlockingEvent(Event):
+    """A special event that blocks the execution of Workflow run.
+
+    To unblock, the user will have to send an object of the class defined in
+    unblocking_event_type.
+    """
 
-class HumanResponseEvent(Event):
+    unblocking_event_type: ClassVar[Type[UnblockingEvent]]
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/workflow/events.py:154]: Why do we need to know the unblocking type? Shouldn't that already be known by the workflow

i.e. a step will emit one event, and accept another, with the assumption that something external is acting as the step

File: llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py [hunk 2]
```diff
@@ -53,6 +53,20 @@ def draw_all_possible_flows(
                 shape="ellipse",
             )  # Light green for events
 
+        for return_type in step_config.return_types:
+            print(f"{return_type}", flush=True)
+            if return_type == type(None):
+                continue
+
+            if issubclass(return_type, BlockingEvent):
```
Comment (logan-markewich) [inline llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py:61]: I wonder if we should insert some generic "external event" into the diagram to connect the blocking and unblocking events

File: llama-index-core/llama_index/core/workflow/events.py (hunk 1 above)
Comment (nerdai) [reply llama-index-core/llama_index/core/workflow/events.py:154]: Yea I think you're probably right it can be inferred by the workflow. That is, if you have a step that produces a blocking event then the next step's accepted event is the unblocking event? Guess I might be worried about some weird edge cases?

I make use of this coupling in my current PR for llama-deploy tho:

```
            async for ev in handler.stream_events():
                # check if blocking event
                if isinstance(ev, BlockingEvent):

                    # wait for the unblocking event type
                    logger.debug(f"Waiting for unblocking event: {ev}")
                    unblocking_ev = await self._events_buffer[current_call.task_id][
                        ev.unblocking_event_type
                    ].get()
```

Comment (logan-markewich) [reply llama-index-core/llama_index/core/workflow/events.py:154]: I think you don't even need to assume that -- the developer of the workflow could code it such that other work is done while waiting for the human response/unblocking event

Basically, the only purpose for having (Un)BlockingEvent is to indicate to the workflow to skip validation for those events, because the actual step that consumes/emits those events is external to the workflow

File: llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py (hunk 2 above)
Comment (logan-markewich) [reply llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py:61]: I think it would be an external step

So like, any blocking event would be linked to some (grey?) step labelled "External Step", and that step emits any unblocking event

Just thought it might make the generated diagram look nicer

Comment (nerdai) [reply llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py:61]: image

I was in the process of doing this and chose "grey-ish" color for it haha. Happy to make changes to the style if we're still not happy with it.

File: llama-index-core/llama_index/core/workflow/events.py (hunk 1 above)
Comment (nerdai) [reply llama-index-core/llama_index/core/workflow/events.py:154]: That makes sense. I probably just need to re-think my implementation a little on the llama-deploy side.

File: llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py (hunk 2 above)
Comment (logan-markewich) [reply llama-index-utils/llama-index-utils-workflow/llama_index/utils/workflow/draw.py:61]: Yea that looks pretty good!

File: llama-index-core/llama_index/core/workflow/events.py (hunk 1 above)
Comment (nerdai) [reply llama-index-core/llama_index/core/workflow/events.py]: alright, removed coupling as its no longer needed on llama-deploy or in drawing :)

File: llama-index-utils/llama-index-utils-workflow/pyproject.toml
Comment (nerdai) [inline llama-index-utils/llama-index-utils-workflow/pyproject.toml:1]: @logan-markewich I guess I should put this in another PR after the `BlockingEvent` and `UnblockingEvent` gets merged and the latest version of core gets released.

**Review by logan-markewich — APPROVED** [review]

**Comment by logan-markewich** [comment]
Looks like deeplake update 😓 Safe to ignore those tests

**Comment by nerdai** [comment]
> Looks like deeplake update 😓 Safe to ignore those tests

🥲 okay will do. But what about coveralls check? I looked at the tests that we currently have and it wasn't immediately obvious to me how we could test the produced `net` to see if it has correctly added the external step node and its relevant edges.

If we deem this as really important, I can spend some time tinkering tests perhaps, tho I don't yet know how would I go about that.

**Comment by logan-markewich** [comment]
Eh, tests are less important for this drawing package (pretty hard to test I agree) -- any test would just generate the png but you can't really check it, only ensure that it didn't break
